File: app/src/main/java/com/example/criminalintent/database/CrimeCursorWrapper.java
package com.example.criminalintent.database;

import android.database.Cursor;
import android.database.CursorWrapper;
import java.util.Date;
import java.util.UUID;

public class CrimeCursorWrapper extends CursorWrapper {
    public CrimeCursorWrapper(Cursor cursor) {
        super(cursor);
    }

    public com.example.criminalintent.Crime getCrime() {
        String uuidString = getString(getColumnIndexOrThrow(CrimeDbSchema.CrimeTable.Cols.UUID));
        String title = getString(getColumnIndexOrThrow(CrimeDbSchema.CrimeTable.Cols.TITLE));
        long date = getLong(getColumnIndexOrThrow(CrimeDbSchema.CrimeTable.Cols.DATE));
        int isSolved = getInt(getColumnIndexOrThrow(CrimeDbSchema.CrimeTable.Cols.SOLVED));
        int suspectIndex = getColumnIndex(CrimeDbSchema.CrimeTable.Cols.SUSPECT);
        int suspectPhoneNumberIndex = getColumnIndex(CrimeDbSchema.CrimeTable.Cols.SUSPECT_PHONE_NUMBER);
        int requiresPoliceIndex = getColumnIndex(CrimeDbSchema.CrimeTable.Cols.REQUIRES_POLICE);
        String suspect = suspectIndex >= 0 ? getString(suspectIndex) : null;
        String suspectPhoneNumber = suspectPhoneNumberIndex >= 0 ? getString(suspectPhoneNumberIndex) : null;
        boolean requiresPolice = requiresPoliceIndex >= 0 && getInt(requiresPoliceIndex) != 0;

        com.example.criminalintent.Crime crime =
                new com.example.criminalintent.Crime(UUID.fromString(uuidString));
        crime.setTitle(title);
        crime.setDate(new Date(date));
        crime.setSolved(isSolved != 0);
        crime.setSuspect(suspect);
        crime.setSuspectPhoneNumber(suspectPhoneNumber);
        crime.setRequiresPolice(requiresPolice);

        return crime;
    }
}

File: app/src/main/java/com/bignerdranch/android/criminalintent/database/CrimeCursorWrapper.kt
package com.bignerdranch.android.criminalintent.database

import android.database.Cursor
import android.database.CursorWrapper
import com.bignerdranch.android.criminalintent.Crime
import com.bignerdranch.android.criminalintent.database.CrimeDbSchema.CrimeTable
import java.util.Date
import java.util.UUID

class CrimeCursorWrapper(cursor: Cursor) : CursorWrapper(cursor) {

    fun getCrime(): Crime {
        val uuidString = getString(getColumnIndex(CrimeTable.Cols.UUID))
        val title = getString(getColumnIndex(CrimeTable.Cols.TITLE))
        val date = getLong(getColumnIndex(CrimeTable.Cols.DATE))
        val isSolved = getInt(getColumnIndex(CrimeTable.Cols.SOLVED))
        val suspect = getString(getColumnIndex(CrimeTable.Cols.SUSPECT))
        val requiresPolice = getInt(getColumnIndex(CrimeTable.Cols.REQUIRES_POLICE))

        return Crime(UUID.fromString(uuidString)).apply {
            this.title = title
            this.date = Date(date)
            this.isSolved = isSolved != 0
            this.suspect = suspect
            this.requiresPolice = requiresPolice != 0
        }
    }
}
